# Árvores II - Tarefa I

Neste exercício vamos trabalhar com a base de dados de identificação de atividade humana com smartphones. Smartphones possuem acelerômetro e giroscópio, que registram dados de aceleração e giro nos eixos X, Y e Z, com frequencia de 50 Hz (ou seja, 50 registros por segundo). Os dados foram agrupados a medidas de 0.3 Hz, calculando-se variáveis derivadas como aceleração mínima, máxima, média etc por eixo no período agrupado de tempo, resultando em 561 variáveis que podem ser lidas nas bases disponíveis.

A base é oriunda de um experimento, em que os indivíduos realizavam uma de seis atividades corriqueiras:
- andando
- subindo escada
- descendo escada
- parado
- sentado
- deitado

O objetivo é classificar a atividade humana com base nos dados do acelerômetro e giroscópio do celular.

### 1. Carregar a base

Sua primeira atividade é carregar a base.

Ela está disponível neste link:
https://archive.ics.uci.edu/ml/datasets/human+activity+recognition+using+smartphones

dados https://archive.ics.uci.edu/ml/machine-learning-databases/00240/  
dataset.names https://archive.ics.uci.edu/ml/machine-learning-databases/00240/UCI%20HAR%20Dataset.names  
UCI HAR Dataset.zip https://archive.ics.uci.edu/ml/machine-learning-databases/00240/UCI%20HAR%20Dataset.zip

Você vai encontrar os seguintes arquivos:

- **features.txt:** contém a lista com os nomes das variáveis
- **features_info.txt:** contém a descrição das variáveis
- **README.txt:** contém uma descrição do estudo e das bases
- **activity_labels:** contém o código da atividade (inteiro entre 1 e 6) e a descrição
- **subject_train.txt:** uma lista indicando que registro pertence a que indivíduo na base de treino
- **X_train.txt:** as *features* (ou variáveis explicativas) da base de testes. Cada linha representa um registro das informações de um indivíduo em um intervalo de tempo de aproximadamente 1/0.3 segundo. As medidas estão em ordem temporal dentro do estudo, e o indivíduo que originou a medida está identificado na base subject_train.txt.
- **y_train.txt:** Possui o mesmo número de linhas que X_test. Contém um número de 1 a 6 indicando a atividade que estava sendo realizada por aquele registro na base de treino.
- **subject_test.txt:** uma lista indicando que registro pertence a que indivíduo na base de teste
- **X_test.txt:** as *features* (ou variáveis explicativas) da base de testes. Cada linha representa um registro das informações de um indivíduo em um intervalo de tempo de aproximadamente 1/0.3 segundo. As medidas estão em ordem temporal dentro do estudo, e o indivíduo que originou a medida está identificado na base subject_test.txt.
- **y_test.txt:** Possui o mesmo número de linhas que X_train. Contém um número de 1 a 6 indicando a atividade que estava sendo realizada por aquele registro na base de teste.

Carregue as bases:

1. Faça o download dos arquivos.
2. Carregue a base ```features.txt``` em uma *Series* (por exemplo usando o comando ```pd.read_csv()```.
3. Carregue a base subject_train.txt em uma *Series*
4. Carregue a base X_train.txt
    1. Faça com que as colunas deste *dataframe* tenham os nomes indicados em ```features.txt```
    2. Sem alterar a ordem dos *dataframes*, coloque o indicador do indivíduo lido em ```subject_train.txt``` como uma variável a mais neste *dataframe***
    3. Faça com que este *dataframe* tenha um índice duplo, composto pela ordem dos dados e pelo identificador do indivíduo
5. Com a mesma lógica, carregue a base X_test
    1. Certifique-se de que tenha os nomes lidos em ```features.txt```
    2. Coloque o identificador do sujeito lido em ```subject_test.txt```
    3. Defina um índice duplo composto pela ordem do registro e o identificador do registro
6. Salve as bases em arquivos CSV para facilitar a leitura deles na terefa 2 deste módulo
7. Considere que esta base é maior que a da aula, tanto em linhas quanto em colunas. Selecione apenas as três primeiras colunas da base ('tBodyAcc-mean()-X', 'tBodyAcc-mean()-Y' e 'tBodyAcc-mean()-Z'), para efeitos desse exercício.

**OBS:** As bases já estão divididas em treino e teste, de modo que não vamos precisar da função ```train_test_split```.

In [142]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
import numpy as np
import matplotlib.pyplot as plt

In [143]:
# Carregando o features.txt
features = pd.read_csv("/content/features.txt", sep=" ", header=None, usecols=[1])

In [144]:
# Carregando o subject_train.txt
subject_train = pd.read_csv("/content/subject_train.txt", header=None)

## Carregando e preparando a base de treino.

In [145]:
# Carregando o X_train e y_train
X_train = pd.read_csv("/content/X_train.txt", on_bad_lines="skip", header=None, delim_whitespace=True)
y_train = pd.read_csv("/content/y_train.txt", on_bad_lines="skip", header=None, delim_whitespace=True)

In [146]:
y_train.columns = ["activity"]

In [147]:
# Alterando o nome das colunas de X_train para os dados de features
X_train.columns = features[1]

In [148]:
# Adicionando nova coluna com o indicador do indivíduo
X_train["indicator"] = subject_train[0]

In [149]:
# Alterando o index para um index duplo
X_train = X_train.reset_index().set_index(["index", "indicator"])

In [150]:
X_train.head(2)

1                tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  \
index indicator                                                            
0     1                   0.288585          -0.020294          -0.132905   
1     1                   0.278419          -0.016411          -0.123520   

1                tBodyAcc-std()-X  tBodyAcc-std()-Y  tBodyAcc-std()-Z  \
index indicator                                                         
0     1                 -0.995279         -0.983111         -0.913526   
1     1                 -0.998245         -0.975300         -0.960322   

1                tBodyAcc-mad()-X  tBodyAcc-mad()-Y  tBodyAcc-mad()-Z  \
index indicator                                                         
0     1                 -0.995112         -0.983185         -0.923527   
1     1                 -0.998807         -0.974914         -0.957686   

1                tBodyAcc-max()-X  ...  fBodyBodyGyroJerkMag-meanFreq()  \
index indicator                    ...                                    
0     1                 -0.934724  ...                        -0.074323   
1     1                 -0.943068  ...                         0.158075   

1                fBodyBodyGyroJerkMag-skewness()  \
index indicator                                    
0     1                                -0.298676   
1     1                                -0.595051   

1                fBodyBodyGyroJerkMag-kurtosis()  angle(tBodyAccMean,gravity)  \
index indicator                                                                 
0     1                                -0.710304                    -0.112754   
1     1                                -0.861499                     0.053477   

1                angle(tBodyAccJerkMean),gravityMean)  \
index indicator                                         
0     1                                      0.030400   
1     1                                     -0.007435   

1                angle(tBodyGyroMean,gravityMean)  \
index indicator                                     
0     1                                 -0.464761   
1     1                                 -0.732626   

1                angle(tBodyGyroJerkMean,gravityMean)  angle(X,gravityMean)  \
index indicator                                                               
0     1                                     -0.018446             -0.841247   
1     1                                      0.703511             -0.844788   

1                angle(Y,gravityMean)  angle(Z,gravityMean)  
index indicator                                              
0     1                      0.179941             -0.058627  
1     1                      0.180289             -0.054317  

[2 rows x 561 columns]

# Carregando e preparando a base de testes.

In [151]:
# Carregando o X_test e y_test
X_test = pd.read_csv("/content/X_test.txt", header=None, delim_whitespace=True, on_bad_lines="skip")
y_test = pd.read_csv("/content/y_test.txt", on_bad_lines="skip", header=None, delim_whitespace=True)

In [152]:
y_test.columns = ["activity"]

In [153]:
# Alterando o nome das colunas para os dados de features
X_test.columns = features[1]

In [154]:
# Adicionando uma nova coluna com o indicador do indivíduo
X_test["indicator"] = subject_train[0]

In [155]:
# Alterando o index para um index duplo
X_test = X_test.reset_index().set_index(["index", "indicator"])

In [156]:
X_test.head(2)

1                tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  \
index indicator                                                            
0     1                   0.257178          -0.023285          -0.014654   
1     1                   0.286027          -0.013163          -0.119083   

1                tBodyAcc-std()-X  tBodyAcc-std()-Y  tBodyAcc-std()-Z  \
index indicator                                                         
0     1                 -0.938404         -0.920091         -0.667683   
1     1                 -0.975415         -0.967458         -0.944958   

1                tBodyAcc-mad()-X  tBodyAcc-mad()-Y  tBodyAcc-mad()-Z  \
index indicator                                                         
0     1                 -0.952501         -0.925249         -0.674302   
1     1                 -0.986799         -0.968401         -0.945823   

1                tBodyAcc-max()-X  ...  fBodyBodyGyroJerkMag-meanFreq()  \
index indicator                    ...                                    
0     1                 -0.894088  ...                         0.071645   
1     1                 -0.894088  ...                        -0.401189   

1                fBodyBodyGyroJerkMag-skewness()  \
index indicator                                    
0     1                                -0.330370   
1     1                                -0.121845   

1                fBodyBodyGyroJerkMag-kurtosis()  angle(tBodyAccMean,gravity)  \
index indicator                                                                 
0     1                                -0.705974                     0.006462   
1     1                                -0.594944                    -0.083495   

1                angle(tBodyAccJerkMean),gravityMean)  \
index indicator                                         
0     1                                       0.16292   
1     1                                       0.01750   

1                angle(tBodyGyroMean,gravityMean)  \
index indicator                                     
0     1                                 -0.825886   
1     1                                 -0.434375   

1                angle(tBodyGyroJerkMean,gravityMean)  angle(X,gravityMean)  \
index indicator                                                               
0     1                                      0.271151             -0.720009   
1     1                                      0.920593             -0.698091   

1                angle(Y,gravityMean)  angle(Z,gravityMean)  
index indicator                                              
0     1                      0.276801             -0.057978  
1     1                      0.281343             -0.083898  

[2 rows x 561 columns]

### 2. Ajustar uma árvore de decisão

- 2.1 ajuste uma árvore de decisão com número mínimo de observações por folha = 20. Certifique-se de que você esteja utilizando apenas as 3 variáveis indicadas no exercício anterior.
- 2.2 Calcule os ccp_alphas. Como feito em aula, certifique-se de que todos os valores são positivos, e selecione apenas valores únicos.
- 2.3 Considere que vamos fazer uma árvore para cada valor de ```ccp_alpha```. Para ter um pouco mais de velocidade, crie uma coleção de dados com os ```ccp_alphas```, considerando apenas 1 a cada 5 valores. Dica: utilize o slicing do tipo ```array[::5]``` para isto. Caso se sinta seguro, fique à vontade para utilizar mais valores de ```ccp_alpha```.

In [157]:
%%time

# Ajustando o modelo
tree = DecisionTreeClassifier(min_samples_leaf=20, random_state=42)
tree.fit(X_train, y_train)

CPU times: user 4.52 s, sys: 8.66 ms, total: 4.52 s
Wall time: 4.54 s


DecisionTreeClassifier(min_samples_leaf=20, random_state=42)

In [158]:
# Calculando o ccp_alpha
path = tree.cost_complexity_pruning_path(X_train, y_train)
ccp_alphas, impurities = path.ccp_alphas, path.impurities

In [159]:
ccp_alphas[::5]

array([0.00000000e+00, 1.33132630e-05, 4.20417450e-05, 1.18184358e-04,
       2.83929989e-04, 4.36662789e-04, 8.77587644e-04, 1.58565659e-03,
       1.93571957e-03, 2.23027206e-03, 3.44050914e-03, 9.92305929e-03,
       9.08568244e-02])

### 3. Desempenho da árvore por ccp_alpha

- 3.1: Rode uma árvore para cada ```ccp_alpha```, salvando cada árvore em uma lista
- 3.2: Calcule a acurácia de cada árvore na base de treinamento e de teste
- 3.3: Monte um gráfico da acurácia em função do ```ccp_alpha``` na base de validação e na base de teste
- 3.4: Selecione a melhor árvore como sendo aquela que possui melhor acurácia na base de teste
- 3.5: Qual a melhor acurácia que se pode obter com estas três variáveis?

**Dica:** utilize a estrutura do notebook apresentado em aula.  
**Dica 2:** meça o tempo com a função mágica ```%%time``` na primeira linha da célula.  
**Sugestão:** caso fique confortável com o tempo de execução, faça a busca pelo melhor ```ccp_alpha``` com mais iterações.  
**Sugestão 2:** caso fique confortável com o tempo de execução, tente inserir uma ou mais variáveis adicionais e veja se consegue aumentar a acurácia.

In [85]:
accuracy_by_ccp = []

In [166]:
%%time
# Criando uma árvore para cada ccp_alpha, calculando sua acurácia e criando um dicionário com a acurácia e o ccpl_alpha
for ccp_alpha in ccp_alphas[::5]:
  tree = DecisionTreeClassifier(min_samples_leaf=20, random_state=42, ccp_alpha=ccp_alpha).fit(X_train, y_train)
  y_pred = tree.predict(X_test)
  accuracy = accuracy_score(y_test, y_pred)

  accuracy_by_ccp.append({
      "ccp_alpha": ccp_alpha,
      "accuracy": accuracy
  })

CPU times: user 58.1 s, sys: 54.6 ms, total: 58.1 s
Wall time: 1min 1s


In [172]:
import plotly.express as px

In [170]:
dataframe = pd.DataFrame(accuracy_by_ccp)

In [181]:
fig = px.line(
    x=dataframe["ccp_alpha"],
    y=dataframe["accuracy"],
    labels={"x": "ccp_alpha", "y": "accuracy"}
)

In [182]:
fig.show()

In [189]:
dataframe.loc[dataframe["accuracy"].idxmax()]

ccp_alpha    0.001936
accuracy     0.865965
Name: 8, dtype: float64

### O melhor ccp_alpha é o 0.001936
- Nota-se que quanto maior ccp_alpha, menor é a acurácia